test/test_ops.py

```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
             for sample in samples:
                 op(sample.input, *sample.args, **sample.kwargs)
 
+    # Helper for comparing torch tensors and numpy arrays
+    def assertEqualHelper(self, actual, expected, msg, *, dtype, exact_dtype=True, **kwargs):
+        assert isinstance(actual, torch.Tensor)
+
+        # Some NumPy functions return scalars, not arrays
+        if isinstance(expected, Number):
+            self.assertEqual(actual.item(), expected, **kwargs)
+        elif isinstance(expected, np.ndarray):
+            # Handles exact dtype comparisons between arrays and tensors
+            if exact_dtype:
+                # Allows array dtype to be float32 when comparing with bfloat16 tensors
+                # since Numpy doesn't support the bfloat16 dtype
+                # Also ops like scipy.special.erf, scipy.special.erfc etc. promote float16
+                # to float32
+                if expected.dtype == np.float32:
+                    assert actual.dtype in (torch.float16, torch.bfloat16, torch.float32)
+                else:
+                    assert expected.dtype == torch_to_numpy_dtype_dict[actual.dtype]
+
+            self.assertEqual(actual,
+                             torch.from_numpy(expected).to(actual.dtype),
+                             msg,
+                             exact_device=False,
+                             **kwargs)
+        else:
+            self.assertEqual(actual, expected, msg, exact_device=False, **kwargs)
+
+    def _test_reference_numerics(self, dtype, op, tensors, equal_nan=True):
+        def _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan=True):
+            if not torch.can_cast(numpy_to_torch_dtype_dict[expected.dtype.type], dtype):
+                exact_dtype = False
+
+            if dtype in [torch.uint8, torch.int8, torch.bool]:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                       exact_dtype=exact_dtype, rtol=1e-3, atol=1e-2)
+
+            elif dtype is torch.bfloat16:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                       exact_dtype=exact_dtype, rtol=16e-3, atol=1e-5)
+
+            else:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
+
+        for sample_input in tensors:
```

Comment:
Looks like the purpose of this block is to convert a sample input to a valid NumPy input, right? Instead of putting this logic here, what about extending the SampleInput class with a `.numpy()` method? It's OK if that method throws an error if an argument is a container for now.

Reply:
Having a `.numpy()` method will be neat, I agree. I was just curious if SciPy supports NumPy arrays as inputs, and looks like it does. (tested for a few ops in SciPy py passing `np.ndarray` input).

I'm sorry though, could you explain a little on what you meant by "if an argument is a container for now"? Do you mean we should only convert to a NumPy array if the input is a torch tensor or a scalar, and not if it's a list/tuple like containers? @mruberry

Reply:
SciPy will also accept `torch.Tensor` instances, as long as they live on CPU and don't use autograd (SciPy will convert it to numpy.ndarray in its input parsing).

"container" is typically a list/tuple/set/etc.

Reply:
Thanks, @rgommers! Appreciate the help and clarification.